services/storage/src/simcore_service_storage/simcore_s3_dsm.py

```diff
@@ -523,22 +527,19 @@ async def delete_file(
                 if not can.delete:
                     raise FileAccessRightError(access_right="delete", file_id=file_id)
 
-        with suppress(FileMetaDataNotFoundError):
-            # NOTE: deleting might be slow, so better ensure we release the connection
-            async with self.engine.acquire() as conn:
-                file: FileMetaDataAtDB = await db_file_meta_data.get(
-                    conn, TypeAdapter(SimcoreS3FileID).validate_python(file_id)
-                )
-            await get_s3_client(self.app).delete_objects_recursively(
-                bucket=file.bucket_name,
-                prefix=(
-                    ensure_ends_with(file.file_id, "/")
-                    if file.is_directory
-                    else file.file_id
-                ),
-            )
-            async with self.engine.acquire() as conn:
-                await db_file_meta_data.delete(conn, [file.file_id])
+            enclosing_file = await find_enclosing_file(conn, user_id, file_id)
+            if enclosing_file:
+                if enclosing_file.file_id == file_id:
+                    await db_file_meta_data.delete(conn, [file_id])
+                else:
+                    await db_file_meta_data.upsert(
+                        conn, await self._update_database_from_storage(enclosing_file)
+                    )
+
```

Comment:
here I have 2 concerns.
1. you now delete the entry in the database **before** actually deleting it in S3 --> that means if anything happens while deleting in S3, then the DB is out of sync. That is bad and may lead to other issues
2. the upsert call is now blocking a potentially very long time, as the update here will re-compute the size of the directory.Another option is to make the row in the DB "invalid". For that you just need to set the file_size entry to -1. See `is_file_entry_valid` entry in services/storage/src/simcore_service_storage/utils.py

Now if I may I would propose the following:
can't you use this function: `list_filter_with_partial_file_id` ? we know the user AND the project_id AND the node_id as the file_id are constructed such as `project_id/node_id/my/amazing/number/of/subfolders/and/file` and it looks like that function should return directories as well

then you check if it is a directory, if yes you just invalidate the size in the row (the next call to get file will re-compute the total size lazily)
if it is a file you delete it, *after* the call to S3.

Reply:
Thanks for the tips, many things were not clear just reading the code.